Add documentation to enable and configure K3s (#10)
Cray-HPE/docs-uan: docs/portal/developer-portal/advanced/Enabling_K3s.md

# Configuring a UAN for K3s (Technical Preview)

**WARNING**: This feature is currently a Technical Preview, as such it requires completion of a prerequisites section. More functionality related to streamlining the configuration and enhancing the experience of rootless podman containers will be added in a future release. This means some of the configuration options may be expected to change.

## UAI Experience on UANs
In UAN 2.6, a new playbook has been added to create a single node, K3s cluster. The K3s environment will be used to run services necessary to replicate the experience of User Access Instances, UAIs, on one or more UANs.

### Use of K3s
K3s will serve as the orchestrator of services necessary to replicate the capabilities of UAIs on UAN hardware. This includes HAProxy, MetalLB, and eventually DNS services like ExternalDNS and PowerDNS. Notably, this does **not** orchestrate instances of sshd and podman containers through K3s. As a comparison, K3s and the initial set of services are operating in a similar way to "Broker UAIs" in CSM to handle the SSH ingress and redirection of users into their interactive environment.

> ### Use of Podman

Author: The podman containers will run with `--net=host` so they will have external access and be able to run jobs.

Traditional UAIs in CSM required some level of privilege in CSM for access to host volume mounts, networking, and startup activities. Podman offers an attractive solution for an interactive environment in which to place users, they can be rootless containers that do not rely on privilege escalation. When running on UANs, podman containers have access to a hosting environment that is already tailored to users.

### Overview
The overall component flow for replicating containerized environments for End-Users on UANs is as follows:

> - A user uses `ssh` to initiate a connection the HAProxy load-balancer running in K3s.

Reviewer: "connection the HAProxy" missing a "to".  Should be "connection to the HAProxy"

- HAProxy, using the configured load balancing algorithms, will forward the SSH connection to an instance of `sshd`
- `sshd`, running on a UAN via systemd, will initiate a rootless podman container as the user using the `ForceCommand` configuration.
- The user is placed in a podman container for an interactive session, or their `SSH_ORIGINAL_COMMAND` is run in the container.

> - When the user disconnects, the podman process exits, and the container is removed.

Author: I'll change the list format and finish that sentence with a period, but there is limitation on where the SSH connection is forwarded to. It could be the same UAN, a different one, or not a UAN at all.

> - When the user disconnects, the podman process exits, and the container is removed.

Author: On second thought, I think your suggestion makes sense as the very next line mentions sshd running on a UAN! This document should be more narrowly focussed on that use case.

There are alternate configurations of podman that would allow for different workflows, for example, the main pid of the container could be long running, to facilitate easier reentry to the container on subsequent logins. 

## Prerequisites 

> The following sections should be completed prior to configuring the UAN with K3s.
> 
> ### UAN K3s Node
> Designate a UAN to operate as the K3s control-plane node.
> 
> **Note**: In a future release, additional UANs will be able to join as extra manager or worker nodes. 
> 
> ### MetalLB IP Address Pool
> 
> A pool of IP Address should be configured for services running in K3s. This will allow for external `LoadBalancer` IP Address to be assigned to services like `HAProxy`.  This address pool must be routable from the UAN control-plane, and should be unused for other purposes. Initially, these IP address will serve as the SSH ingress for instances of `HAProxy`.
> 
> ### Subuid and Subgid
> 
> To allow for users to run rootless podman containers, the files `/etc/subuid` and `/etc/subgid` must be present and configured with an entry for each user. These files should be uploaded to the `user` S3 bucket:
> ```bash
> $ cray artifacts list user --format json
> {
>   "artifacts": [
>     {
>       "Key": "subuid",
>       "LastModified": "2023-02-21T23:41:43.948000+00:00",
>       "ETag": "\"c543aebb9b40bcf48879885734447090\"",
>       "Size": 145686,
>       "StorageClass": "STANDARD",
>       "Owner": {
>         "DisplayName": "User Service User",
>         "ID": "USER"
>       }
>     },
>     {
>       "Key": "subgid",
>       "LastModified": "2023-02-21T23:41:43.948000+00:00",
>       "ETag": "\"73032ede132e44d2c1bc567246901737\"",
>       "Size": 145686,
>       "StorageClass": "STANDARD",
>       "Owner": {
>         "DisplayName": "User Service User",
>         "ID": "USER"
>       }
>     }
>   ]
> }
> ```

Reviewer 2: ```suggestion
The following steps should be completed prior to configuring the UAN with K3s.

1. Designate a UAN to operate as the K3s control-plane node.

   **Note**: In a future release, additional UANs will be able to join as extra manager or worker nodes. 

1. Identify a pool of IP addresses for the services running in K3s.
   
   This address pool must be routable from the UAN control-plane, and should be unused for other purposes.

   This will allow for external `LoadBalancer` IP Address to be assigned to services like `HAProxy`. Initially, these IP addresses will serve as the SSH ingress for instances of `HAProxy`.

1. Configure and create the `/etc/subuid` and `/etc/subgid` files.

   To allow for users to run rootless podman containers, these files must be present and configured with an entry for each user. These files should be uploaded to the `user` S3 bucket:

   ```bash
   $ cray artifacts list user --format json
   {
     "artifacts": [
       {
         "Key": "subuid",
         "LastModified": "2023-02-21T23:41:43.948000+00:00",
         "ETag": "\"c543aebb9b40bcf48879885734447090\"",
         "Size": 145686,
         "StorageClass": "STANDARD",
         "Owner": {
           "DisplayName": "User Service User",
           "ID": "USER"
        }
       },
       {
         "Key": "subgid",
         "LastModified": "2023-02-21T23:41:43.948000+00:00",
         "ETag": "\"73032ede132e44d2c1bc567246901737\"",
         "Size": 145686,
         "StorageClass": "STANDARD",
         "Owner": {
           "DisplayName": "User Service User",
           "ID": "USER"
         }
       }
     ]
   }
   ```
```

### Podman Image

> A container image suitable for users should be available in a container image registry accessible from the UAN.

Reviewer 2: This should also be converted into a step and part of the same ordered list as the above steps:

1. Ensure that a container image suitable for users is available in a container image registry accessible from the UAN.

## Configuring with Configuration Framework Service (CFS)

> After completing the prerequisites section, the following should be established to proceed with configuring a UAN to run K3s.

Reviewer 2: ```suggestion
After completing the [Prerequisites](#prerequisites) section, the following are available to proceed with configuring a UAN to run K3s.
```

- A fully configured UAN
- An IPAddress start and end range to assign to MetalLB
- Prepared subuid and subgid files 

### Configuration Files and Playbook

Configuration for K3s and related services are controlled in the following ansible files in the UAN VCS repository:
```bash
$ ls uan-config-management/vars/ | egrep "k3s|sshd|helm"
uan_helm.yml
uan_k3s.yml
uan_sshd.yml
```
The ansible playbook to install and configure K3s may be found here:
```bash
$ ls uan-config-management | grep k3s
k3s.yml
```

### Ansible Roles

> The following Ansible roles are provided in the `uan-config-management` repository in VCS. There are `README.md` files in each Ansible role directory in this repository that provide further details.

Reviewer 2: Should we update the "Ansible Roles" file in the documentation now for these new roles, or wait until this is more of a stable, supported feature?

Author: I would prefer to wait. Also, I'd like to come up with a way to extract those READMEs programmatically instead of duplicating the information as it inevitably becomes out of sync.

- **uan-k3s-install**: Download and stage K3s assets necessary to initialize and configure k3s on a node
- **uan-k3s-stage**: Install and configure K3s on a node
- **uan-helm**: Perform tasks to initialize an environment to install helm charts
- **uan-haproxy**: Deploy a list of HAProxy charts to the k3s cluster
- **uan-metallb**: Deploy the MetalLB chart to the k3s cluster
- **uan-sshd**: Create and enable instances of `sshd` with systemd

### Artifactory Assets

UAN uploads artifacts to deploy to the UAN control-plane node in a new nexus repository:

- uan-2.6.XX-third-party

The following Nexus group repository is created and reference the aforementioned UAN Nexus raw repos.

- uan-2.6-third-party

This repository will contain the installer for K3s, Helm charts for HAProxy and MetalLB, etc.

### Validation Tests 

To validate the K3s cluster once deployed, see the [Validation Checks](#validation-checks) section of this document for details.

> ## Getting Started

Reviewer 2: ```suggestion
## Configure UANs for User Podman Containers
```

> Perform the following tasks to get it going! todo fix this

Reviewer: This ok to be here with the "todo fix this"

Author: Whoops, certainly not.

Author: Looks like I missed a couple of "todo" lines. I'll address all of them and push up a new commit.

> Perform the following tasks to get it going! todo fix this

Reviewer 2: ```suggestion
```

> ### MetalLB 

Reviewer 2: This and the following Level 3 sections should be converted into a single ordered list since they depict a sequence of steps (and their explanations and examples), similar to what I did above with the prerequisites.

Author: I'm hesitant to do this as this is not a 100% prescriptive step-by-step set of instructions. Would it be acceptable to start with a note that each of the sections should be read and understood to determine how a site may want to configure each of the components?

Reviewer 2: There are some optional steps, supporting info, and substeps. I think it would still be better to refactor this as a procedure. What if I do the change and use this web GUI to suggest the change? Then you can accept/reject it.

Reviewer 2: Since this is a Technical Preview feature, we can leave it as it is in this commit, with the explanatory paragraph at the top saying that these sections need to be read in full before being performed. I'll procedure-fy this later.

- **uan-haproxy**: Deploy a list of HAProxy charts to the k3s cluster
- **uan-metallb**: Deploy the MetalLB chart to the k3s cluster
- **uan-sshd**: Create and enable instances of `sshd` with systemd

### Artifactory Assets

UAN uploads artifacts to deploy to the UAN control-plane node in a new nexus repository:

- uan-2.6.XX-third-party

The following Nexus group repository is created and reference the aforementioned UAN Nexus raw repos.

- uan-2.6-third-party

This repository will contain the installer for K3s, Helm charts for HAProxy and MetalLB, etc.

### Validation Tests 

To validate the K3s cluster once deployed, see the [Validation Checks](#validation-checks) section of this document for details.

## Configuring K3s, MetalLB, HAProxy, and SSHD for use with Podman

Each of the sections below describe how the various comonents deployed to K3s and the UANs may be configured to enable users to SSH to rootless podman containers. As there is no one configuration to fit any one use case, read and understand each section to modify the configuration as needed. Once each section has been completed, see [Deploy K3s to the UAN](#deploy-k3s-to-the-uan).

### MetalLB 

> Configure the start and end range for MetalLB `IPAddressPool` in `vars/uan_helm.yml`:

Reviewer 2: Can we offer any hard requirements or guidance about the bare minimum number of IPs that need to be in the metallb address pool? Once  the customer knows that, then they'll quickly decide whether it's worth the effort to reclaim some IPs from CSM.

Author: Good suggestion.

```bash
# kubectl get services -n ims | grep ims
# kubectl get services -n user | grep uai
# kubectl get services -n uai | grep 
```

It may be possible to reallocate the CSM MetalLB pool `customer-access` from CSM to make room for a subset of IPs to use with MetalLB on UANs. To shrink the `customer-access` pool in CSM, edit the configmap and pick a new CIDR block for for `customer-access`. In this example the CIDR block was `x.x.x.x/26`:
```bash
# kubectl edit -n metallb-system cm/metallb
...
data:
  config: |
    address-pools:
    - addresses:
      - x.x.x.x/27
      name: customer-access
      protocol: bgp
...
```
This leaves a portion of IP Address unallocated that may then be used to set `metallb_ipaddresspool_range_start` and `metallb_ipaddresspool_end`. 

**Important**: When calculating the range of IP Address now available from `customer-access`. Be sure to account for the Broadcast IP of the remaining `customer-access` pool.

To complete migrating the IP Address range out of CSM, restart the MetalLB controller pod in CSM.
```bash
# kubectl delete pod -n metallb-system -l app.kubernetes.io/component=controller
```

### HAProxy Configuration

> Each SSH ingress into is backed by a K3s deployment of HAProxy. By default, a single instance of HAProxy is enabled in `vars/uan_helm.yml`:

Reviewer 2: missing word between "into" and "is" in the first sentence. I think "UAN" should go there.

    chart: "{{ haproxy_chart }}"
    chart_path: "{{ helm_install_path }}/charts/{{ haproxy_chart }}.tgz"
    args: "--set service.type=LoadBalancer"
    config: |
      global
        log stdout format raw local0
        maxconn 1024
      defaults
        log     global
        mode    tcp
        timeout connect 10s
        timeout client 36h
        timeout server 36h
        option  dontlognull
      listen ssh
        bind *:22
        balance leastconn
        mode tcp
        option tcp-check
        tcp-check expect rstring SSH-2.0-OpenSSH.*
        server uan01 uan01.example.domain.com:9000 check inter 10s fall 2 rise 1
        server uan02 uan02.example.domain.com:9000 check inter 10s fall 2 rise 1
        server uan03 uan03.example.domain.com:9000 check inter 10s fall 2 rise 1
```
This is an example that should be tailored to the desired configuration. See the [SSHD Configuration](#sshd-configuration) section to create new instances of SSHD to respond to HAProxy connections outside of the standard SSHD running on port 22.

For more information HAProxy configurations, see [HAProxy Configuration](#https://docs.haproxy.org/2.7/configuration.html).

> To additional instances of HAProxy representing alternate configurations, add a new element to the list `uan_haproxy`.

Reviewer 2: missing word (probably "enable") between "To" and "additional".

metallb_ipaddresspool_range_end: "x.x.x.x"
```

> MetalLB will assign an IP address to each service running in K3s that requires and external IP address. In the case of HAProxy, each instance of HAProxy will require an IP address. Podman containers do *not* require their own IP address.

Reviewer 2: in the future when this feature because more supported and mature, we can use this fact as the start of the decision tree  which will help me procedure-fy this and the following config sections.
